File: src/lib/domain/epicScope.ts
import type { LocalTask, Tray } from "../types";

export const TBD_EPIC_SCOPE = "TBD";

export type EpicScopeConfig = Pick<Tray, "epicScope" | "transversalEpicScope">;

export type EpicTarget = {
  project: string;
  area: string;
  scope: string;
};

export function normalizeEpicScope(scope?: string | null): string | null {
  const normalized = scope?.trim().replace(/\s+/g, " ");
  if (!normalized) return null;
  if (normalized.toLowerCase() === "[tbd]" || normalized.toLowerCase() === "tbd") return TBD_EPIC_SCOPE;
  return normalized;
}

export function epicTargetForTask(
  task: Pick<LocalTask, "project" | "area">,
  config?: EpicScopeConfig
): EpicTarget | null {
  const scope = effectiveEpicScopeForProject(task.project, config);
  if (!scope) return null;
  return {
    project: task.project.trim(),
    area: task.area.trim(),
    scope
  };
}

export function effectiveEpicScopeForProject(project: string, config?: EpicScopeConfig): string | null {
  const singular = normalizeEpicScope(config?.epicScope);
  if (!singular) return null;
  if (singular === TBD_EPIC_SCOPE) return singular;
  if (project.trim().toLowerCase() === "transversal") {
    return normalizeEpicScope(config?.transversalEpicScope);
  }
  return singular;
}

export function formatEpicTarget(target: EpicTarget): string {
  return `[${target.project}] [${target.area}] ${target.scope}`;
}

export function formatPendingEpicTarget(project: string, area: string): string {
  return `[${project.trim()}] [${area.trim()}] Scope pending`;
}

export function suggestTransversalEpicScope(scope?: string | null): string | null {
  const normalized = normalizeEpicScope(scope);
  if (!normalized) return null;
  if (normalized === TBD_EPIC_SCOPE) return TBD_EPIC_SCOPE;

  const words = normalized.split(" ");
  const firstWord = words[0] ?? "";
  const lowerFirstWord = firstWord.toLowerCase();
  if (lowerFirstWord.endsWith("s")) return normalized;
  if (lowerFirstWord.endsWith("ión")) {
    words[0] = `${firstWord.slice(0, -3)}iones`;
    return words.join(" ");
  }
  if (lowerFirstWord.endsWith("z")) {
    words[0] = `${firstWord.slice(0, -1)}ces`;
    return words.join(" ");
  }
  if (/[aeiouáéíóú]$/i.test(firstWord)) {
    words[0] = `${firstWord}s`;
    return words.join(" ");
  }
  words[0] = `${firstWord}es`;
  return words.join(" ");
}

File: src-tauri/src/epic_scope.rs
use crate::models::LocalTask;

pub(crate) const TBD_SCOPE: &str = "TBD";

#[derive(Debug, Clone, Copy)]
pub(crate) struct EpicScopeConfig<'a> {
    pub(crate) singular_scope: Option<&'a str>,
    pub(crate) transversal_scope: Option<&'a str>,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct EpicTarget {
    pub(crate) local_project: String,
    pub(crate) area: String,
    pub(crate) scope: String,
}

impl EpicTarget {
    pub(crate) fn for_task(task: &LocalTask, config: EpicScopeConfig<'_>) -> Option<Self> {
        Some(Self {
            local_project: task.project.trim().to_string(),
            area: task.area.trim().to_string(),
            scope: effective_scope_for_project(&task.project, config)?,
        })
    }

    pub(crate) fn summary(&self) -> String {
        format!("[{}] [{}] {}", self.local_project, self.area, self.scope)
    }

    pub(crate) fn legacy_summary(&self) -> String {
        format!("[{}] {}", self.local_project, self.area)
    }

    pub(crate) fn searchable_summaries(&self) -> [String; 2] {
        [self.summary(), self.legacy_summary()]
    }

    pub(crate) fn matches_existing_summary(&self, summary: &str) -> bool {
        summary == self.summary() || summary == self.legacy_summary()
    }
}

pub(crate) fn effective_scope_for_project(
    project: &str,
    config: EpicScopeConfig<'_>,
) -> Option<String> {
    let singular = normalize_epic_scope(config.singular_scope)?;
    if singular == TBD_SCOPE {
        return Some(singular);
    }
    if project.trim().eq_ignore_ascii_case("Transversal") {
        return normalize_epic_scope(config.transversal_scope);
    }
    Some(singular)
}

pub(crate) fn normalize_epic_scope(scope: Option<&str>) -> Option<String> {
    let normalized = scope?.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.is_empty() {
        return None;
    }
    if normalized.eq_ignore_ascii_case("[TBD]") || normalized.eq_ignore_ascii_case(TBD_SCOPE) {
        return Some(TBD_SCOPE.to_string());
    }
    Some(normalized)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_tbd_without_brackets() {
        assert_eq!(normalize_epic_scope(Some("[TBD]")).as_deref(), Some("TBD"));
        assert_eq!(normalize_epic_scope(Some(" tbd ")).as_deref(), Some("TBD"));
    }

    #[test]
    fn resolves_transversal_scope_only_when_confirmed() {
        let config = EpicScopeConfig {
            singular_scope: Some("Demo Version 1"),
            transversal_scope: None,
        };

        assert_eq!(
            effective_scope_for_project("STT", config).as_deref(),
            Some("Demo Version 1")
        );
        assert_eq!(effective_scope_for_project("Transversal", config), None);
    }

    #[test]
    fn tbd_scope_applies_to_transversal_without_plural_flow() {
        let config = EpicScopeConfig {
            singular_scope: Some("TBD"),
            transversal_scope: None,
        };

        assert_eq!(
            effective_scope_for_project("Transversal", config).as_deref(),
            Some("TBD")
        );
    }

    #[test]
    fn epic_target_owns_current_and_legacy_summaries() {
        let target = EpicTarget {
            local_project: "STT".to_string(),
            area: "Programacion".to_string(),
            scope: "Demo Version 1".to_string(),
        };

        assert_eq!(target.summary(), "[STT] [Programacion] Demo Version 1");
        assert_eq!(
            target.searchable_summaries(),
            [
                "[STT] [Programacion] Demo Version 1".to_string(),
                "[STT] Programacion".to_string(),
            ]
        );
        assert!(target.matches_existing_summary("[STT] [Programacion] Demo Version 1"));
        assert!(target.matches_existing_summary("[STT] Programacion"));
    }
}
